composer/core/evaluator.py

```diff
@@ -59,6 +61,44 @@ def should_eval(state: State, event: Event):
             last_batch_seen = state.timestamp.batch
             return True
 
+        if eval_interval.unit == TimeUnit.DURATION:
+            assert state.max_duration is not None, 'max_duration should not be None'
+            if state.dataloader_len is None:
+                raise RuntimeError(
+                    f'Evaluation interval of type `dur` or {TimeUnit.DURATION} requires the dataloader to be sized.')
+            if state.max_duration.unit == TimeUnit.EPOCH and int(
+                    state.timestamp.batch) % math.ceil(state.max_duration.value * float(eval_interval) *
+                                                       state.dataloader_len) == 0 and event == Event.BATCH_END:
+                last_batch_seen = state.timestamp.batch
+                return True
+            elif state.max_duration.unit == TimeUnit.BATCH and int(state.timestamp.batch) % math.ceil(
+                    state.max_duration.value * eval_interval.value) == 0 and event == Event.BATCH_END:
+                last_batch_seen = state.timestamp.batch
+                return True
+            elif (state.max_duration.unit == TimeUnit.SAMPLE or
+                  state.max_duration.unit == TimeUnit.TOKEN) and event == Event.BATCH_END:
+                # Scenarios:
+                # (1) If the last global sample in a current batch is evenly divisible (zero remainder) by a evaluation interval duration
+                # during each evaluation cycle.
+                # E.g.: If batch_size = 2, max_duration.value = 50, eval_interval = 0.2
+                #       then, last sample in a batch: 2, 4, 6, 8, 10, 12, 14, 16, 18, 20, 22,......., 50
+                #       eval interval duration (theoretically): 10, 20, 30, 40, 50
+                #       eval interval duration (practically):   10, 20, 30, 40, 50
+                # (2) If the last global sample in a current batch is NOT evenly divisible (zero remainder) by a evaluation interval duration
+                # during each evaluation cycle, then it performs evaluation during first batch of next evaluation cycle.
+                # E.g.: If batch_size = 2, max_duration.value = 50, eval_interval = 0.1
+                #       then, last sample in a batch: 2, 4, 6, 8, 10, 12, 14, 16, 18, 20, 22,......., 50
+                #       eval interval duration (theoretically): 5, 10, 15, 20, 25, 30, 35, 40, 45, 50
+                #       eval interval duration (practically):   6, 10, 16, 20, 26, 30, 36, 40, 46, 50
+                # Logic: It should perform evaluation if the co-efficient of last global sample in a current batch by evaluation interval
+                #        duration does not match with last global sample in a previous batch by evaluation interval duration.
```

Comment:
```suggestion
                # If last sample in batch is not evenly divisible by eval_interval, perform evaluation in next batch
```
We generally want to avoid long comments in the code. While this is very valuable for me to understand it in a PR review, I don't think we need to give examples and detailed explanations. The code speaks for itself, and our style guide dictates minimal comments. Lets just leave a small comment for the edge case in scenario 2

Reply:
Cool. Done in https://github.com/mosaicml/composer/pull/1523/commits/24a6f7ea7f073bc06805508ce7f9d2205fd97d46.